yt/frontends/ramses/particle_handlers.py

```diff
@@ -0,0 +1,329 @@
+
+    See `SinkParticleFileHandler` for an example implementation.'''
+
+    # These properties are static properties
+    ptype = None  # The name to give to the particle type
+    fname = None  # The name of the file(s).
+    file_descriptor = None # The name of the file descriptor (if any)
+
+    attrs = None  # The attributes of the header
+    known_fields = None  # A list of tuple containing the field name and its type
+
+    # These properties are computed dynamically
+    field_offsets = None     # Mapping from field to offset in file
+    field_types = None       # Mapping from field to the type of the data (float, integer, ...)
+    local_particle_count = None  # The number of particle in the domain
+
+    def __init__(self, ds, domain_id):
+        '''
+        Initalize an instance of the class. This automatically sets
+        the full path to the file. This is not intended to be
+        overriden in most cases.
+
+        If you need more flexibility, rewrite this function to your
+        need in the inherited class.
+        '''
+        self.ds = ds
+        self.domain_id = domain_id
+        basename = os.path.abspath(
+              os.path.dirname(ds.parameter_filename))
+        iout = int(
+            os.path.basename(ds.parameter_filename)
+            .split(".")[0].
+            split("_")[1])
+        icpu = domain_id
+
+        self.fname = os.path.join(
+            basename,
+            self.fname.format(iout=iout, icpu=icpu))
+
+        if self.file_descriptor is not None:
+            self.file_descriptor = os.path.join(
+                basename,
+                self.file_descriptor)
+
+    @property
+    def exists(self):
+        '''
+        This function should return True if the *file* the instance
+        exists. It is called for each file of the type found on the
+        disk.
+
+        By default, it just returns whether the file exists. Override
+        it for more complex cases.
+        '''
+        return os.path.exists(self.fname)
+
+    @property
+    def has_part_descriptor(self):
+        '''
+        This function should return True if a *file descriptor*
+        exists.
+
+        By default, it just returns whether the file exists. Override
+        it for more complex cases.
+        '''
+        return os.path.exists(self.file_descriptor)
+
+    @classmethod
+    def any_exist(cls, ds):
+        '''
+        This function should return True if the kind of particle
+        represented by the class exists in the dataset. It takes as
+        argument the class itself -not an instance- and a dataset.
+
+        Arguments
+        ---------
+        * ds: a Ramses Dataset
+
+        Note
+        ----
+        This function is usually called once at the initialization of
+        the RAMSES Dataset structure to determine if the particle type
+        (e.g. regular particles) exists.
+        '''
+        raise NotImplementedError
```

Comment:
I know you note that these need to be implemented by subclasses in the class docstring, but for people who miss that can you add a comment like `# this function must be implemented by subclasses` to this function and the other functions that raise `NotImplementedError` in this pull request.